src/components/SignUp/SignUp.tsx

```diff
@@ -0,0 +1,156 @@
+import { useState, useEffect } from 'react';
+import "./SignUp.css";
+import { authorize, register } from '../../utils/auth';
+import { setToken } from '../../utils/token';
+import { getUserInfo } from '../../utils/api';
+import { useNavigate, Link } from 'react-router-dom';
+import { useDispatch, useSelector } from 'react-redux';
+import { RootState } from '../../redux/store';
+import { setUser } from '../../redux/userSlice';
+import Form from '../Form/Form';
+
+type FormData = {
+  name: string;
+  email: string;
+  password: string;
+  avatar?: string;
+  bio?: string;
+};
+
+// const defaultAvatarPath = '../../assets/avatar-placeholder.jpg';
+
+function SingUp() {
+  const navigate = useNavigate();
+  const dispatch = useDispatch();
+  const user = useSelector((state: RootState) => state.user);
+  const [isLoading, setIsLoading] = useState(false);
+  const [data, setData] = useState({
+    name: '',
+    email: '',
+    password: '',
+    avatar: '',
+    bio: '',
+  });
+
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
```

Comment:
If it’s interesting for you here is a custom `validation hook`.

You should put it in a separate file  `useFormAndValidation.js` in folder `hooks` and then import it to the component that needs validation:

```js

import {useState, useCallback} from 'react';

export function useFormAndValidation() {
  const [ values, setValues ] = useState({});
  const [ errors, setErrors ] = useState({});
  const [ isValid, setIsValid ] = useState(true);

  const handleChange = (e) => {
    const {name, value} = e.target
    setValues({...values, [name]: value });
    setErrors({...errors, [name]: e.target.validationMessage});
    setIsValid(e.target.closest('form').checkValidity());
  };



  const resetForm = useCallback((newValues = {}, newErrors = {}, newIsValid = false) => {
    setValues(newValues);
    setErrors(newErrors);
    setIsValid(newIsValid);
  }, [setValues, setErrors, setIsValid]);

  return { values, handleChange, errors, isValid, resetForm, setValues, setIsValid };
}


```

And only one line of code enables validation:

```js

  const {values, handleChange, errors, isValid, setValues, resetForm} = useFormAndValidation()

```

So, you **don’t need** to make functions for handling inputs, storing values, resetting the form. Everything is already done.